### PR TITLE
Add coursier implementation of dependency resolution API

### PR DESCRIPTION
**Pull request by leonardehrenfried**
This implements a Coursier implementation of the dependency management API.

This exploratory implementation is somewhat beta as some functionality is missing. I am opening this pull request in order to gauge interest in the wider community in this approach and whether it's worth me spending time polishing this further.

With only this PR, sbt itself will not include a way of using this. A separate PR on sbt/sbt is needed for this.

## General approach

I've tried to use as much as possible from the coursier plugin itself. In particular the classes `FromSbt` and `ToSbt` are straight up copies from their repository.

## Not implemented yet/problems

- Classifiers
- Plugins are successfully resolved but don't show up in the sbt command prompt. How does plugin auto-discovery work?
- Coursier does not implement cached resolution
- Coursier does not implement timing reports in UpdateReport

I am hoping that @dwijnand, @eed3si9n and @alexarchambault can provide feedback on how to implement the missing features.

## Questions

- Should the local Ivy cache still be included as a default resolver?
- What is the meaning of ~~UpdateReport~~ `ModuleDescriptor.extraInputHash`?

As I said above I would love some feedback about whether this has a chance of being merged or if there are any blockers that need to be removed first.

### References
- https://github.com/coursier/coursier/issues/601
- https://github.com/sbt/sbt/issues/2997

### CODE REVIEW

**Comment by typesafe-tools** [comment]
To validate this pull request against other sbt modules, please visit <A HREF="https://jenkins.scala-sbt.org:8592/job/sbt-validator/parambuild/?librarymanagement=pull/190/head&deploy_to_bintray=false&run_sbt_community=false">this link.</A>

**Comment by alexarchambault** [comment]
@leonardehrenfried I can't review it in detail right now, but that looks really cool.

About the ivy2 cache, if you meant `~/.ivy2/local`, yes it should be kept, as coursier doesn't handle publishing for now. If you meant `~/.ivy2/cache`, it shouldn't, but that one shouldn't show up, does it?

Is the code from `FromSbt` and `ToSbt` simple copy / paste? Maybe it could be spin up to a separate `coursier-sbt` module (with mima checks) in the sources of coursier.

**Comment by leonardehrenfried** [comment]
> About the ivy2 cache, if you meant ~/.ivy2/local, yes it should be kept, as coursier doesn't handle publishing for now. If you meant ~/.ivy2/cache, it shouldn't, but that one shouldn't show up, does it?

I was thinking about both. Right now the Ivy dep resolution includes them by default. I would also think that `local` makes a lot of sense but not sure if `cache` should be kept for backwards-compatibility. I added it manually but happy to remove it again.

> Is the code from FromSbt and ToSbt simple copy / paste? Maybe it could be spin up to a separate coursier-sbt module (with mima checks) in the sources of coursier.

Yes, it's intentionally a dumb c&p so sharing code is made easier. I would love a reusable module that I can depend on!

**Comment by eed3si9n** [comment]
@leonardehrenfried Thanks for taking on this task!

> What's the meaning of extraInputHash?

`extraInputHash` is part of `ModuleDescriptor`, not `UpdateReport`.
Since `update` is needed for classpaths, and classpaths are needed for `compile`, etc `update` is called all the time, and the result is normally cached. This means that we need to be able to detect when we need to actually perform the `update` task (resolving and downloading) when things change on the `ModuleDescriptor`. Since `ModuleDescriptor` is an open trait, it won't know about fields its subclasses might introduce. This is where `extraInputHash` is used.

For Ivy it looks like this:

```scala
    def extraInputHash: Long = {
      import AltLibraryManagementCodec._
      Hasher.hash(owner.configuration) match {
        case Success(keyHash) => keyHash.toLong
        case Failure(_)       => 0L
      }
    }
```

**Comment by eed3si9n** [comment]
If we are talking about bringing Coursier into to the librarymanagement code base (as opposed to living in the plugin space as alternative lm) one of my concerns about Coursier is its dependencies (see https://github.com/coursier/coursier/issues/175, https://github.com/coursier/coursier/issues/25, https://github.com/lihaoyi/Ammonite/issues/644, etc).

@alexarchambault Do you have rough timeline for 1.0 and 1.1?

**Comment by leonardehrenfried** [comment]
Coursier 1.0.0 has been released today, so I will update the PR accordingly. Handily, this includes a module that contains all the sbt <-> coursier conversion, so the diff will become smaller.

@eed3si9n I understand your reservation about pulling in Scalaz. In [this comment](https://github.com/lihaoyi/Ammonite/issues/644#issuecomment-310349651) @alexarchambault says that he will be working on removing the scalaz dependency. Do you feel it's condition of this PR being merged or is it simply a strong preference?

**Comment by eed3si9n** [comment]
Since Alex has been publicly on board with reducing the dependencies, I'd much rather wait until it's done.

**Comment by alexarchambault** [comment]
@eed3si9n Yep, it's something I'd like to tackle quickly.

**Comment by jvican** [comment]
Is there any progress on this awesome effort?

**Comment by leonardehrenfried** [comment]
There is _some_ progress, but there are a couple of issues that block me from progressing further:

- @eed3si9n doesn't want to merge this in until the Scalaz dependency is removed. @alexarchambault said that this will happen "soon".
- coursier 1.0.0 has a separate module for the sbt<->coursier translations. That's very useful but that also depends on sbt itself creating a circular dependency. This led to a weird compilation/linking problem, but I can't remember the details right now.
- Luckily it can be avoided (I think) by moving some of the 1.0/0.13 compatibility code from coursier to @dwijnand's sbt-compat: https://github.com/dwijnand/sbt-compat/pull/5 . ~Now Dale only has to release this*, so I can remove the offending code from coursier and reduce the dependency to librarymanagement, not all of sbt.~ (Already done: https://github.com/coursier/coursier/pull/731) 

So in summary, there are a few things that need to fall into place for this to happen. If people want to help out they could work on removing Scalaz from coursier.

*Excellent argument for sbt-release-early.

**Comment by leonardehrenfried** [comment]
For future reference, here is the compilation/linking error that I'm getting with coursier 1.0.0:

```
[error] /home/travis/build/sbt/librarymanagement/coursier/src/main/scala/sbt/librarymanagement/coursier/CoursierDependencyResolution.scala:68:39: Symbol 'type <none>.Import.Resolver' is missing from the classpath.
[error] This symbol is required by 'value coursier.FromSbt.resolver'.
[error] Make sure that type Resolver is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
[error] A full rebuild may help if 'FromSbt.class' was compiled against an incompatible version of <none>.Import.
[error]       reorderedResolvers.flatMap(r => FromSbt.repository(r, ivyConfiguration, log, authentication)) ++ Seq(
[error]                                       ^
[error] /home/travis/build/sbt/librarymanagement/coursier/src/main/scala/sbt/librarymanagement/coursier/CoursierDependencyResolution.scala:145:9: Symbol 'type <none>.Import.Logger' is missing from the classpath.
[error] This symbol is required by 'value coursier.ToSbt.log'.
[error] Make sure that type Logger is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
[error] A full rebuild may help if 'ToSbt.class' was compiled against an incompatible version of <none>.Import.
[error]         ToSbt.updateReport(
[error]         ^  
```

As I said upthread I think it's a problem with `coursier-shared` depending on sbt which messes up the classpath, but I'm having a hard time figuring out what exactly this module depends on. I will take another look once 1.0.1 is released.

**Comment by leonardehrenfried** [comment]
After having taken a look here https://repo1.maven.org/maven2/io/get-coursier/sbt-shared_2.12_1.0/1.0.0/sbt-shared-1.0.0.pom

I see that `coursier-shared` definitely depends on `"org.scala-sbt" %% "sbt"`. That should probably be reduced to just `"org.scala-sbt" %% "sbt-librarymanagement"`.

I'll raise ticket with coursier.

**Comment by leonardehrenfried** [comment]
Now that coursier 1.0.1-M1 has been released I rebased this branch and tried to compile it. Unfortunately, it cannot resolve `sbt-compat` because it's not on Maven Central. Should it be able to find it? (https://travis-ci.org/sbt/librarymanagement/jobs/336468744#L724)

Besides, I was under the impression that `sbt-compat` is a compile-time-only dependency of `coursier-shared` and doesn't need to be resolved by `lmCoursier`.

On the other hand, my dependency declaration on sbt-coursier is a bit strange, but I couldn't get it working otherwise: https://github.com/sbt/librarymanagement/pull/190/files#diff-0ecdbc5a001d52fb34f5eafb7cd1aaa6R51

(The artifact I want to depend on is https://repo1.maven.org/maven2/io/get-coursier/sbt-shared_2.12_1.0/1.0.1-M1/sbt-shared-1.0.1-M1.pom)

@dwijnand Can you advise?

**Comment by dwijnand** [comment]
sbt-compat is only released to Bintray: repo.scala-sbt.org/scalasbt/sbt-plugin-releases/com.dwijnand/sbt-compat/

**Comment by leonardehrenfried** [comment]
Is it OK to add that to the resolvers?

**Comment by leonardehrenfried** [comment]
I saw this PR mentioned in the Scala Contributor thread so I thought I give a quick update.

After a lot of moving code around coursier's `sbt-shared` and @dwijnand's `sbt-compat` modules, I managed to get this branch to compile. However, it turned out that I missed yet another instance of types from `sbt._` which the unit tests showed. The fix has already been merged in coursier's master and I'm awaiting the release of 1.0.3.

The second open issue was the Scalaz dependency. This was discussed in https://github.com/coursier/coursier/issues/175 and @alexarchambault has already started the task of removing Scalaz in https://github.com/coursier/coursier/pull/781. Getting rid of Scalaz is no easy task so I think this will take a little longer.

**Comment by leonardehrenfried** [comment]
coursier 1.1.0-M1 has been released a few days ago and I have adjusted to the new API.

Notably, @alexarchambault has worked hard on removing the scalaz dependency in coursier-core and cache, so this is now scalaz-free!

I think that warrants another round of review and commenting.

One question: Travis currently only fails on Scala 2.11 because `coursier/sbt-shared` isn't being released for 2.11 (as there is no sbt for 2.11). What is the best way to disable `lmCoursier` for 2.11?

I tried removing 2.11 from `crossVersions` but that led to this error: https://travis-ci.org/sbt/librarymanagement/jobs/354221428#L660

**Comment by eed3si9n** [comment]
I think a good place to start is building sbt from source with this patch, and making sure that the dependency management tests (scripted dependency-management)

**Comment by eed3si9n** [comment]
> One question: Travis currently only fails on Scala 2.11 because coursier/sbt-shared isn't being released for 2.11 (as there is no sbt for 2.11). What is the best way to disable lmCoursier for 2.11?

@dwijnand or I should look into this.

**Comment by eed3si9n** [comment]
2.11 is an artifact from the time sbt 1.x beta was built using 2.11, so I am ok with just getting rid of 2.11 from the lm build altogether. I'd be more interested in 2.13.0-M3 than 2.11 at this point.

**Comment by leonardehrenfried** [comment]
@eed3si9n I think removing 2.11 is no longer necessary as I will have to copy the ToSbt and FromSbt conversion objects from coursier's source code anyway.

For the live of me I cannot get sbt to resolve the transitive sbt plugins of couriser/sbt-shared. I think it has to do with sbt-shared being hosted on Maven Central and not Bintray. :|

**Comment by leonardehrenfried** [comment]
For an example see: https://travis-ci.org/leonardehrenfried/sbt/jobs/355248584#L3974

**Comment by leonardehrenfried** [comment]
Ok, progress report time: I copied ToSbt and FromSbt back into this project and for the first time I can get this branch to compile and all unit tests to pass!

I've created my own branch of sbt proper where I'm working on integrating this into sbt itself.

I've hit a problem though and would appreciate help. I'm seeing this compilation error when running sbt's scripted tests:

```
[info] [info] Compiling 1 Scala source to /tmp/sbt_61fe949b/project/target/scala-2.12/sbt-1.0/classes ...
[info] [error] /tmp/sbt_61fe949b/project/InstrumentScripted.scala:2:8: Symbol 'type sbt.std.TaskExtra' is missing from the classpath.
[info] [error] This symbol is required by 'package sbt.package'.
[info] [error] Make sure that type TaskExtra is in your classpath and check for conflicting dependencies with `-Ylog-classpath`.
[info] [error] A full rebuild may help if 'package.class' was compiled against an incompatible version of sbt.std.
[info] [error] import sbt._, Keys._
[info] [error]        ^
[info] [error[0m] /tmp/sbt_61fe949b/project/InstrumentScripted.scala:2:15: not found: object Keys
[info] [error] import sbt._, Keys._
[info] [error]               ^
[info] [error] /tmp/sbt_61fe949b/project/InstrumentScripted.scala:3:35: not found: type AutoPlugin
[info] [error] object InstrumentScripted extends AutoPlugin {
```

https://travis-ci.org/leonardehrenfried/sbt/jobs/359840184#L1055

Does anyone know what this means?

**Comment by eed3si9n** [comment]
> I've hit a problem though and would appreciate help. I'm seeing this compilation error when running sbt's scripted tests:

I've seen that sometimes during local development, and usually `clean` fixes it. Do you see that locally too?

**Comment by alexarchambault** [comment]
Wouldn't adding coursier as a git submodule be ok, rather than copy-pasting `FromSbt` / `ToSbt`? (manually adding `FromSbt.scala` / `ToSbt.scala` via ~`unmanagedSources`)

**Comment by leonardehrenfried** [comment]
> I've seen that sometimes during local development, and usually clean fixes it. Do you see that locally too?

@eed3si9n I haven't actually run the scripted tests locally but will try.

Should this issue be even possible on CI with a clean `target` folder?

> Wouldn't adding coursier as a git submodule be ok, rather than copy-pasting FromSbt / ToSbt? (manually adding FromSbt.scala / ToSbt.scala via ~unmanagedSources)

@alexarchambault I would be open to this but I'm not sure what the policy for this is. Maybe Eugene or Dale can chime in.

**Comment by eed3si9n** [comment]
sbt's license is https://github.com/sbt/sbt/blob/1.x/LICENSE + https://www.lightbend.com/contribute/cla. This grants copyright to Lightbend, so we probably should get the written ok from @alexarchambault on this. My general preference is that we host librarymanagement code in sbt/librarymanagement repo. I'd also be happy to give repo rights to both of you assuming this PR goes through.

**Comment by eed3si9n** [comment]
> Should this issue be even possible on CI with a clean target folder?

I am not sure exactly. One thing I noticed is that `InstrumentScripted` is not in any package, and there's something odd that happens if you don't put an object in a package, so could you try changing that? https://github.com/sbt/sbt/blob/v1.1.2/scripted/sbt/src/main/scala/sbt/test/ScriptedTests.scala#L136

**Comment by leonardehrenfried** [comment]
@eed3si9n I tried putting `InstrumentScripted` into a package but that didn't work. I also tried it locally and unfortunately it didn't work either.

Can you take a look at the diff of my branch of sbt and check if there is anything in there that could cause the failure: https://github.com/sbt/sbt/compare/1.x...leonardehrenfried:coursier

Here is the latest run of the scripted test of that branch: https://travis-ci.org/leonardehrenfried/sbt/jobs/360044405

**Comment by eed3si9n** [comment]
Something is not hooked up correctly either in your lm or sbt. I added the following line to debug your classpath:

```diff
diff --git a/scripted/sbt/src/main/scala/sbt/scriptedtest/ScriptedTests.scala b/scripted/sbt/src/main/scala/sbt/scriptedtest/ScriptedTests.scala
index dd9de5bc1..7672a65eb 100644
--- a/scripted/sbt/src/main/scala/sbt/scriptedtest/ScriptedTests.scala
+++ b/scripted/sbt/src/main/scala/sbt/scriptedtest/ScriptedTests.scala
@@ -322,6 +322,7 @@ final class ScriptedTests(
           IO.copyDirectory(originalDir, tempTestDir)

           val runTest = () => {
+            IO.write(tempTestDir / "project" / "p.sbt", """logLevel := Level.Debug""")
```

and I am not seeing sbt JARs passed into Zinc:

```
[info] [debug] [zinc] Running cached compiler 72e02473 for Scala compiler version 2.12.4
[info] [debug] [zinc] The Scala compiler is invoked with:
[info] [debug] 	-deprecation
[info] [debug] 	-bootclasspath
[info] [debug] 	/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/sunrsasign.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/classes:/var/folders/hg/2602nfrs2958vnshglyl3srw0000gn/T/sbt_1e8dd449/global/boot/scala-2.12.4/lib/scala-library.jar
[info] [debug] 	-classpath
[info] [debug] 	/private/var/folders/hg/2602nfrs2958vnshglyl3srw0000gn/T/sbt_1e8dd449/project/target/scala-2.12/sbt-1.0/classes
[info] [error] /private/var/folders/hg/2602nfrs2958vnshglyl3srw0000gn/T/sbt_1e8dd449/project/InstrumentScripted.scala:2:8: not found: object sbt
[info] [error] import sbt._, Keys._
[info] [error]        ^
```

compare that with 1.x branch:

```
[info] [debug] Getting org.scala-sbt:compiler-bridge_2.12:1.1.3:compile for Scala 2.12.4
[info] [debug] [zinc] Running cached compiler 5cc1fdcb for Scala compiler version 2.12.4
[info] [debug] [zinc] The Scala compiler is invoked with:
[info] [debug] 	-deprecation
[info] [debug] 	-bootclasspath
[info] [debug] 	/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/resources.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/rt.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/sunrsasign.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/jsse.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/jce.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/charsets.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/lib/jfr.jar:/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk/Contents/Home/jre/classes:/var/folders/hg/2602nfrs2958vnshglyl3srw0000gn/T/sbt_12ef7ae4/global/boot/scala-2.12.4/lib/scala-library.jar
[info] [debug] 	-classpath
[info] [debug] 	/private/var/folders/hg/2602nfrs2958vnshglyl3srw0000gn/T/sbt_12ef7ae4/project/target/scala-2.12/sbt-1.0/classes:/Users/eed3si9n/.ivy2/local/org.scala-sbt/sbt/1.2.0-SNAPSHOT/jars/sbt.jar:
/Users/eed3si9n/.ivy2/local/org.scala-sbt/main_2.12/1.2.0-SNAPSHOT/jars/main_2.12.jar:
/Users/eed3si9n/.ivy2/local/org.scala-sbt/logic_2.12/1.2.0-SNAPSHOT/jars/logic_2.12.jar:
/Users/eed3si9n/.ivy2/local/org.scala-sbt/collections_2.12/1.2.0-SNAPSHOT/jars/collections_2.12.jar:
/Users/eed3si9n/.ivy2/cache/com.eed3si9n/sjson-new-scalajson_2.12/jars/sjson-new-scalajson_2.12-0.8.2.jar:
/Users/eed3si9n/.ivy2/cache/com.eed3si9n/sjson-new-core_2.12/jars/sjson-new-core_2.12-0.8.2.jar:
/Users/eed3si9n/.ivy2/cache/com.eed3si9n/shaded-scalajson_2.12/jars/shaded-scalajson_2.12-1.0.0-M4.jar:
/Users/eed3si9n/.ivy2/cache/org.spire-math/jawn-parser_2.12/jars/jawn-parser_2.12-0.10.4.jar:
/Users/eed3si9n/.ivy2/cache/org.scala-sbt/util-position_2.12/jars/util-position_2.12-1.1.3.jar:
/Users/eed3si9n/.ivy2/cache/org.scala-sbt/util-relation_2.12/jars/util-relation_2.12-1.1.3.jar:
/Users/eed3si9n/.ivy2/local/org.scala-sbt/actions_2.12/1.2.0-SNAPSHOT/jars/actions_2.12.jar:
/Users/eed3si9n/.ivy2/local/org.scala-sbt/completion_2.12/1.2.0-SNAPSHOT/jars/completion_2.12.jar:
/Users/eed3si9n/.ivy2/cache/jline/jline/jars/jline-2.14.6.jar:
/Users/eed3si9n/.ivy2/cache/org.scala-sbt/io_2.12/jars/io_2.12-1.1.4.jar:
/Users/eed3si9n/.ivy2/cache/net.java.dev.jna/jna/jars/jna-4.5.0.jar:
/Users/eed3si9n/.ivy2/cache/net.java.dev.jna/jna-platform/jars/jna-platform-4.5.0.jar:
/Users/eed3si9n/.ivy2/cache/org.scala-sbt/util-control_2.12/jars/util-control_2.12-1.1.3.jar:
/Users/eed3si9n/.ivy2/local/org.scala-sbt/run_2.12/1.2.0-SNAPSHOT/jars/run_2.12.jar:
/Users/eed3si9n/.ivy2/cache/org.scala-sbt/util-logging_2.12/jars/util-logging_2.12-1.1.3.jar:
/Users/eed3si9n/.ivy2/cache/org.scala-sbt/util-interface/jars/util-interface-1.1.3.jar:
/Users/eed3si9n/.ivy2/cache/org.apache.logging.log4j/log4j-api/jars/log4j-api-2.8.1.jar:
/Users/eed3si9n/.ivy2/cache/org.apache.logging.log4j/log4j-core/jars/log4j-core-2.8.1.jar:
....
/Users/eed3si9n/.ivy2/cache/org.scala-sbt/compiler-bridge_2.12/jars/compiler-bridge_2.12-1.1.3.jar
```

**Comment by jvican** [comment]
> sbt's license is sbt/sbt:LICENSE@1.x + lightbend.com/contribute/cla. This grants copyright to Lightbend, so we probably should get the written ok from @alexarchambault on this. My general preference is that we host librarymanagement code in sbt/librarymanagement repo. I'd also be happy to give repo rights to both of you assuming this PR goes through.

Why should @alexarchambault grant any copyright to Lightbend? sbt is currently using a fork of ivy (sbt/ivy) and happily using ivy's Apache license. There should be no problem that `coursier` is used as is.

**Comment by eed3si9n** [comment]
@jvican 

> Why should @alexarchambault grant any copyright to Lightbend? sbt is currently using a fork of ivy (sbt/ivy) and happily using ivy's Apache license. There should be no problem that `coursier` is used as is.

There's a difference in the way software is licensed, binary is used, and how the source is controlled.

If we use Ivy as an example:

1. as you pointed out, we literally forked it, so we control the patch and release cycle.
2. sbt/librarymanagement uses the forked Ivy as binary dependency
3. lm-ivy code is licensed under BSD-3 with Lightbend copyright.

For Coursier, similarly, @alexarchambault should rightfully retain the copyright etc of the Coursier itself. (At this point, I don't see the need to fork Coursier). But if any _source code_ becomes part of this pull request, then we request BSD-3 with Lightbend CLA to be applied for the glue code. Since Leonard is not the original author of FromSbt.scala and ToSbt.scala, I thought we would eventually require written consent of the CLA.

**Comment by jvican** [comment]
> But if any source code becomes part of this pull request, then we request BSD-3 with Lightbend CLA to be applied for the glue code.

Sure, just wanted to emphasize that the glue code should not be checked in the coursier repository, and therefore there's no need to ask Alex to re-license coursier. My opinion as an outsider is that the big benefit of using coursier, aside from the fact that it's way better than ivy, is that the sbt team doesn't need to maintain it -- it only needs to take care of the code that integrates with it, which is already an advantage over maintaining sbt/ivy! Having coursier versioned and licensed as an independent tool makes things much easier :smile:.

**Comment by leonardehrenfried** [comment]
We might be able to sidestep the entire issue by me trying a little harder to include `coursier/sbt-shared`. It is an sbt plugin of the build but also required at runtime, which I guess is confusing to the dependency resolution.
It is not listed as a dependency of `lmCoursier` when you run `publishLocal` so that's why the `sbt/sbt` build cannot find it. Maybe it works if I manually readd it not as a plugin but as a dependency like this:

`"io.get-coursier" % "sbt-shared_2.12_1.0" % "1.1.0-M1"`

The same would then also be needed for `sbt-compat`.

**Comment by leonardehrenfried** [comment]
Ok, I just tried it out and it doesn't work because that would resolve the following:

`https://repo1.maven.org/maven2/io/get-coursier/sbt-shared_2.12_1.0/1.1.0-M1/sbt-shared_2.12_1.0-1.1.0-M1.pom`

When in actual fact the correct URL is:

`https://repo1.maven.org/maven2/io/get-coursier/sbt-shared_2.12_1.0/1.1.0-M1/sbt-shared-1.1.0-M1.pom`

**Comment by leonardehrenfried** [comment]
Maybe I need to add it as an sbt-plugin of lmCoursier _and_ of sbt itself. But I'm not sure what is worse: Copying two classes or trying to get sbt to transitively include an sbt plugin.

**Comment by leonardehrenfried** [comment]
I figured out why the scripted tests break. The dependency `"org.scala-sbt" % "sbt" % "1.1.0" % "provided"` isn't resolved transitively by coursier because of the `provided` scope.

@eed3si9n @alexarchambault I'm not sure what the correct behaviour is here. Should provided dependencies be transitively resolved?

**Comment by eed3si9n** [comment]
See https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html

> This scope is only available on the compilation and test classpath, and is not transitive.

**Comment by eed3si9n** [comment]
https://github.com/sbt/sbt/blob/v1.1.2/main/src/main/scala/sbt/Defaults.scala#L2068-L2072

```scala
      val sbtdeps =
        (sbtDependency in pluginCrossBuild).value.withConfigurations(Some(Provided.name))
      val pluginAdjust =
        if (isPlugin) sbtdeps +: base
        else base
```

The code above adds sbt as a provided dependency to the metabuild's root project directly. So I am not sure why you'd need to resolve it transitively.

**Comment by leonardehrenfried** [comment]
Then I'm officially out of ideas why this doesn't work.

If those jars are manually added to the classpath, how come that the compiler cannot find `sbt.Keys`?

**Comment by alexarchambault** [comment]
(I'm fine with signing the CLA if that's needed or to clarify things.)

**Comment by dwijnand** [comment]
@leonardehrenfried would you be able to join our sbt meeting? Eugene is not available next week Wednesday, so perhaps Wednesday 13 June? We'd love to get a chance to chat with you (and @alexarchambault too if you'd like to join!) about this PR.

Alternatively or in addition: are you going to be at Scala Days NYC?

**Comment by leonardehrenfried** [comment]
Hi, sorry for not following up on this PR.

I'd gladly join a an sbt meeting either at the coming week or the week after that.

Sadly, I was too snowed under at work to attend scala days in Berlin even though it happened quite close to where I live. :|

**Comment by leonardehrenfried** [comment]
So, to answer your question: no, unfortunately I won't be at Scala Days NYC.

**Comment by andreaTP** [comment]
this  can be closed now.

**Comment by eed3si9n** [comment]
Closing this in favor of https://github.com/sbt/librarymanagement/pull/270
